### ISSUE TITLE
Multi-domain support

### ISSUE THREAD

**Issue by hostingnuggets** [roehling/postsrsd#75]
If I understand correctly PostSRSd can only handle one domain. Is this correct? If not can please someone update the documentation to include an example on how to handle a postfix mail server which handles multiple domains.

Nowadays I don't know of any postfix mail server which only hosts one single domain. For example my postfix servers have hundred of domains which would all require SRS.

**Comment by RandieM** [comment]
The same problem/question here. I have done quite some research and I cannot seem to find a SRS solution for shared mail hosting with postfix. It would be really great if postsrsd supported this, especially now that more and more mail servers use SPF.

**Comment by tomsommer** [comment]
+1 

Basically take the domain from the `From` header and use it as SRS_DOMAIN.

**Comment by jseutens** [comment]
+1 , google has a -all in SPF (hotmail is still ~all)

**Comment by dpavlotzky** [comment]
For what it is worth: this is technically not possible. Postsrsd is a TCP-based lookup table for Postfix. From the [Postfix TCP_TABLE man page](http://www.postfix.org/tcp_table.5.html):

>PROTOCOL DESCRIPTION
       The TCP map class implements a very simple protocol: the client sends a
       request, and the server sends one reply.<br/> \<snip><br/>

>REQUEST FORMAT
       Each request specifies a command, a lookup key, and possibly  a  lookup
       result.

       get SPACE key NEWLINE
              Look up data under the specified key.

>REPLY FORMAT
       Each  reply specifies a status code and text. Replies must be no longer
       than 4096 characters including the newline terminator.

       500 SPACE text NEWLINE
              In case of a lookup request, the requested data does not  exist.
              In  case  of  an  update request, the request was rejected.  The
              text describes the nature of the problem.

       400 SPACE text NEWLINE
              This indicates an error condition. The text describes the nature
              of the problem. The client should retry the request later.

       200 SPACE text NEWLINE
              The request was successful. In the case of a lookup request, the
              text contains an encoded version of the requested data.

So basically what this means is that no state is sent to postsrsd. It is impossible for postsrsd to know what the original recipient address was and so has no way of selecting a correct domain to use for rewriting.

This limitation is not inherent to tcp based lookup tables. Lookup tables in general cannot provide what you want because there is no way to retreive the original recipient.

If you really want this you'll have to look elsewhere. Maybe a milter or have a look at the [postfix pipe manual](http://www.postfix.org/pipe.8.html). The environment of the pipe command does contain the original recipient:

>${original_recipient}
                     This macro expands  to  the  complete  recipient  address
                     before any address rewriting or aliasing.

**Comment by RandieM** [comment]
@dpavlotzky, thanks a lot for taking the time to give such a detailed answer. I guess I will have to find a solution elsewhere...

**Comment by roehling** [comment]
Since there is no technical reason to use the forwarder's domain for the SRS rewrite, my favorite solution so far is using a dedicated SRS domain which is independent of any customer domains (e.g. `srs.mailhoster.org`).

**Comment by RandieM** [comment]
@roehling, would you mind elaborating a little on your solution?
When you say "SRS rewrite", what exactly do you mean? Is it the sender rewrite?
Also, in your proposed set-up, where do bounce emails go?
Thanks!

**Comment by roehling** [comment]
The SRS rewrite is the process of turning `user@example.com` into `SRS0+XXXX+ZZ+example.com+user@srs.mailhoster.org`. 

If you need to forward a mail for *any* of your customer's domains, you can simply forward it with the `srs.mailhoster.org` domain; as long as your mail server is authorized to send on behalf of `srs.mailhoster.org`, the SPF validation will succeed. The bounce will still work as expected, since you are handling mails for `srs.mailhoster.org`, and the bounce address is fully contained within the rewritten one.

**Comment by RandieM** [comment]
@roehling thanks a lot for the explanation. This sounds promising, so I will give it a try, as soon as I get some time.

**Comment by novianbiz** [comment]
is it still not possible?

**Comment by mislav-eu** [comment]
Same problem also here. My hostname is e.g. webXX.mydomain.tld and now, since we host multiple domains under the same mail server, SRS rewrites email from tmp-test@second.tld into "SRS0=thk5=O4=second.tld=tmp-test@mydomain.tld" - and when email is sent to recipient server which checks for SPF/TXT records - this fails, because this server is not allowed to send email in name of my primary domain "mydomain.tld" - this domain should have SRS rewritten to the origin sender domain, not output of "postconf -h mydomain".

**Comment by lakester** [comment]
Per the "srs.mailhoster.org" discussion/suggestion:  I'm in the same boat as the rest of the folks.  **It SOUNDS like I need only add whatever my "srs.mailhoster.org" domain is to the SPF record of all the domains I support.  Is this correct?**  FWIW, my mail gateway is MailGun..., so as a result of the SRS rewrite, I see all forwarded traffic submitted via the credentials of the single domain (i.e., more or less the mydomain config val).

**Comment by defcon8** [comment]
For GMail, Return-Path and Header From are needed to be the same domain or DMARC validation will fail and messages being marked as spam. 

Source: https://webmasters.stackexchange.com/questions/76765/sent-emails-pass-spf-and-dkim-but-fail-dmarc-when-received-by-gmail

My mailserver also relais mails for multiple domains and now SRS rewrites all emails to ....@myprimairydomain.tld. Which differs from the original sender domain and makes GMail DMARC check fail.

Any solution this, we need this urgent.

**Comment by defcon8** [comment]
> The SRS rewrite is the process of turning `user@example.com` into `SRS0+XXXX+ZZ+example.com+user@srs.mailhoster.org`.
> 
> If you need to forward a mail for _any_ of your customer's domains, you can simply forward it with the `srs.mailhoster.org` domain; as long as your mail server is authorized to send on behalf of `srs.mailhoster.org`, the SPF validation will succeed. The bounce will still work as expected, since you are handling mails for `srs.mailhoster.org`, and the bounce address is fully contained within the rewritten one.

Not according GMail. "Return-Path:" and "Header From:" domain needed to be the same or DMARC validation will fail and messages being marked as spam.
